source/opt/value_number_table.cpp

```diff
@@ -0,0 +1,97 @@
+// Copyright (c) 2017 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "value_number_table.h"
+
+#include <algorithm>
+
+namespace spvtools {
+namespace opt {
+
+uint32_t ValueNumberTable::GetValueNumber(spvtools::ir::Instruction* inst) {
+  // TODO: Need to implement the substitution of operands by their value number
+  // before hashing.
+  // TODO: Implement a normal form for opcodes that commute like integer
+  // addition.  This will let us know that a+b is the same value as b+a.
```

Comment:
IEEE 754 float addition and multiplication is supposed to be commutative. Still it feels like playing with fire.

Reply:
Yes, for floating point arithmetic we will have to be careful.  I would expect that we can only do this if we have "relaxed precision" or the like.  We'll have to cross that bridge when we come to it.

Reply:
Forgive the drive-by comment:

For graphics APIs, the default is "fast relaxed" math: assume you can reassociate, commute, and all sorts of numerically unstable things. That can be turned off if there is a "NoContraction" decoration.

For compute APIs (OpenCL), the default is the same as C: can't reassociate or contract, etc.  The rules can be relaxed by applying "FP Fast Math" flags on basic operations, including the Fast and AllowRecip flags.